File: day8.hs
import qualified Data.Map as Map
import qualified Data.Set as Set


type AntiNode = (Int, Int)
type Antenna = (Int, Int)
type AntennaGroup = [Antenna]
type AntennaGroups = Map.Map Char AntennaGroup

getAntenna :: Int -> Char -> Int -> AntennaGroups
getAntenna _ '.' _ = Map.empty
getAntenna y typ x = Map.singleton typ [(x, y)]

antennasInRow :: String -> Int -> AntennaGroups
antennasInRow row rowNum = Map.unionsWith (++) $ zipWith (getAntenna rowNum) row [0..]

cartesUnique :: Eq a => [a] -> [a] -> [(a, a)]
cartesUnique [] _ = []
cartesUnique (x : xs) y = filter (uncurry (/=)) $ map (x,) y ++ cartesUnique xs y

getAntiNodePair :: Antenna -> Antenna -> Set.Set AntiNode
getAntiNodePair a b = Set.fromList [node1, node2]
    where
        x1 = fst a
        x2 = fst b
        y1 = snd a
        y2 = snd b
        dx = x1 - x2
        dy = y1 - y2
        node1 = (x1 + dx, y1 + dy)
        node2 = (x2 - dx, y2 - dy)

inBounds :: Int -> Int -> AntiNode -> Bool
inBounds max_x max_y (x, y) = 0 <= x && x < max_x && 0 <= y && y < max_y

getAntiNodes :: Int -> Int -> AntennaGroup -> Set.Set AntiNode
getAntiNodes width height group = Set.filter (inBounds width height) $ Set.unions $ map (uncurry getAntiNodePair) $ cartesUnique group group

solve :: String -> Int
solve input = Set.size $ Set.unions $ Map.elems $ Map.map (getAntiNodes width height) antennaGroups
    where
        width = length $ head $ lines input
        height = length $ lines input
        antennaGroups = Map.unionsWith (++) $ zipWith antennasInRow (lines input) [0..]

getAntiNodeLine :: Int -> Int -> Antenna -> Antenna -> Set.Set AntiNode
getAntiNodeLine max_x max_y (x1, y1) (x2, y2) = Set.fromList $ takeWhile (inBounds max_x max_y) line1 ++ takeWhile (inBounds max_x max_y) line2
    where
        dx = x1 - x2
        dy = y1 - y2
        line1 = map (\n -> (x1 + dx * n, y1 + dy * n)) [0..]
        line2 = map (\n -> (x2 - dx * n, y2 - dy * n)) [0..]

getAntiNodes2 :: Int -> Int -> AntennaGroup -> Set.Set AntiNode
getAntiNodes2 width height group = Set.filter (inBounds width height) $ Set.unions $ map (uncurry (getAntiNodeLine width height)) $ cartesUnique group group

solve2 :: String -> Int
solve2 input = Set.size $ Set.unions $ Map.elems $ Map.map (getAntiNodes2 width height) antennaGroups
    where
        width = length $ head $ lines input
        height = length $ lines input
        antennaGroups = Map.unionsWith (++) $ zipWith antennasInRow (lines input) [0..]

main :: IO ()
main = do
    input <- readFile "day8.txt"
    let result = solve input
    print result
    let result2 = solve2 input
    print result2

File: day8.py
import aoc
from functools import partial
from itertools import product


def handle_input(raw: str):
    antennas = {}
    lines = raw.splitlines()
    for y in range(len(lines)):
        for x in range(len(lines[y])):
            if lines[y][x] != '.':
                typ = lines[y][x]
                if typ in antennas:
                    antennas[typ].append((x, y))
                else:
                    antennas[typ] = [(x, y)]
    return antennas, (len(lines[0]), len(lines))


def uncurry(func):
    return lambda tup: func(*tup)


def get_anti_nodes(antennas: dict):
    anti_nodes = set()
    for group in antennas.values():
        for node1, node2 in product(group, repeat=2):
            if node1 == node2:
                continue
            dx = node1[0] - node2[0]
            dy = node1[1] - node2[1]
            anti_nodes.add((node1[0] + dx, node1[1] + dy))
            anti_nodes.add((node2[0] - dx, node2[1] - dy))
    return anti_nodes


def in_bounds(max_x, max_y, x, y):
    is_x = 0 <= x < max_x
    is_y = 0 <= y < max_y
    return is_x and is_y


def solve(antennas, size):
    return len(list(filter(uncurry(partial(in_bounds, *size)), get_anti_nodes(antennas))))


def get_line(x, y, dx, dy, size):
    nodes = set()
    for dx, dy in ((dx, dy), (-dx, -dy)):
        for step in range(100000):
            new_node = x + dx * step, y + dy * step
            if in_bounds(*size, *new_node):
                nodes.add(new_node)
            else:
                break
    return nodes


def get_anti_nodes2(antennas: dict, size):
    anti_nodes = set()
    for group in antennas.values():
        for node1, node2 in product(group, repeat=2):
            if node1 == node2:
                continue
            dx = node1[0] - node2[0]
            dy = node1[1] - node2[1]
            anti_nodes |= get_line(*node1, dx, dy, size)
    return anti_nodes


def solve2(antennas, size):
    return len(get_anti_nodes2(antennas, size))


if __name__ == '__main__':
    antennas, size = handle_input(aoc.get_input('day8.txt'))
    res = solve(antennas, size)
    print(res)
    res = solve2(antennas, size)
    print(res)